### ISSUE TITLE
v2.0 roadmap, API overhaul, and more!

### ISSUE THREAD

**Issue by CosmicHorrorDev** [WilliamVenner/steamlocate-rs#7]
Pinging @WilliamVenner since I would love to hear your thoughts on this since it's your library after all :smile:, and @super-continent since you mentioned potentially being interested in helping with some of the changes

## Pending Fixes

- #6
  - `steamy-vdf` parsing issues strike again! This time with app manifests. The obvious way forward is to switch the app manifest parsing to also use `keyvalues-parser`. The issue is that the `SteamApp` type directly exposes a `steamy_vdf::Table`, so this will have to be a breaking change.

### Switching `SteamDir::apps` from a caching API to an iterator based API

The idea here is to switch `SteamDir::apps` from returning a reference to an internal hashmap to returning an iterator that tries parsing each manifest and returning a result for each one. This API is inspired by similar situations like `std::fs::read_dir`.Something like

```rust
impl SteamDir {
    pub fn apps(&self) -> AppIter {
        ...
    }
}

impl Iterator for AppIter {
    // Each entry is the app id followed by a result for reading and parsing the manifest
    type Item = (u32, steamlocate::Result<SteamApp>);

    fn next(&mut self) -> Option<Self::Item> {
        ...
    }
}
```

The reasoning for this switch is that I don't think we gain much by caching, and this switch makes things more flexible along with being more straightforward

### Exposing an error type

The current API just converts all errors that are encountered into `None` which makes it hard to track down the source of issues. The fix here is to enumerate possible error types in an enum and expose it for fallible operations

### Lock down some public internals

Things like `LibraryFolders`' `paths` shouldn't be manipulated by user's of the library, so it should be private with a public getter. The same goes for `SteamDir`'s `path`

### Avoid exposing parsing types in `SteamApp`

It would be good to look over a large set of app manifest files to see how reasonable it would be to expose all the information within the struct instead of having to expose a type from the parser publicly. The reasoning is twofold

1. `keyvalues-parser` is not stable and it's very possible that I will be making breaking changes in the future (I'm pending a breaking release atm). Breaking changes for exposed types would be a breaking change for this library to upgrade to too which isn't ideal
2. Usually messing with the low-level parsed types is gnarly, so it would be much more ergonomic for users to not have to touch it

### Run test in CI

Currently the test suite has a lot of requirements for it to actually be run. It would be good to store some sample libraryfolders and appmanifests, and do some path manipulation to run tests against those instead of requiring steam to be installed with some specific games

### Give `cargo clippy` a go

There's some lints that should be easy enough to cleanup :)

---

If this all sounds good then the task breakdown would be something like

- [x] Swap appmanifest parsing from using `steamy-vdf` to `keyvalues-parser` (or `keyvalues-serde`)
- [x] Switch `SteamDir::apps` to return an iterator for `SteamApp`s
- [x] Expose a custom error type to indicate possible failure cases
- [x] Lock down public fields in `LibraryFolders` and `SteamDir`
- [x] Expose all values for app manifests directly in `SteamApp` instead of exposing the parsed obj
- [x] Revamp tests so that we can run it from within CI
- [x] Fix `cargo clippy` lints
- [ ] Update documentation
- [x] Refactor `App` to be just be the plain data it holds. Move the installation folder resolution to a method on the library
- [ ] Setup a changelog / migration guide
- [x] ~~Test more feature combinations in CI~~ No longer needed now that `locate` is the only feature
- [x] Drop the `locate` feature in favor of target specific dependencies (update related docs)
- [ ] Attempt to detect whether the installation is valid or uninstalled to avoid detecting unused flatpak installs when also installed locally
- [ ] Expose the paths that we check for when searching for linux installations so that people can manually check for installations when desired (e.g. when compiling for wasm, but still running on Linux)

**Comment by WilliamVenner** [comment]
This library could do with a full rewrite really - I wrote it during my first month of learning Rust, so by all means, go to town. I agree with every point here :D

**Comment by medzernik** [comment]
Any progress on this by any chance? :D

**Comment by CosmicHorrorDev** [comment]
Got distracted with other stuff, but I'll work on reviving it

**Comment by CosmicHorrorDev** [comment]
I've been totally swamped lately, but there's the `2.0.0-alpha.0` release published on crates.io that gets rid of `steamy-vdf` (without the rest of these changes)

**Comment by Davidian1024** [comment]
I was looking through the dev-v2.0 branch and it kinda looked like that was the case.  But I wasn't sure.
I'm new enough to rust that I don't know how to use that code in my own project.  Is that pretty straightforward?

**Comment by CosmicHorrorDev** [comment]
API-wise it's largely the same as the current version, aside from changing the fields in the `SteamApp` struct. You would just specify the alpha version in your `Cargo.toml` file

```toml
[dependencies]
steamlocate = "2.0.0-alpha.0"
```

The documentation is also built on docs.rs, but the doc comments themselves haven't been updated, so don't rely on those to be correct

https://docs.rs/steamlocate/2.0.0-alpha.0/steamlocate/steamapp/struct.SteamApp.html

**Comment by Davidian1024** [comment]
Awesome, thank you!

**Comment by Windows10CE** [comment]
Hello, me and @ethangreen-dev are working on https://github.com/thunderstore-io/thunderstore-cli/tree/clingwrap and wanted to ask about a few features that could possibly get included in 2.0 (we'd be happy to implement them ourselves, just want to make sure they'd be welcome here, otherwise we'll probably maintain a fork), namely:
 - A way to test an arbitrary Path to see if its a Steam directory (and by extension build a SteamDir from an arbitrary Path), mostly just so if SteamLocate fails to find Steam (for whatever reason, Linux specifically is very weird we've found) we can ask the user to tell us where Steam is, then locate all its apps from there.
 - Not sure if this is really in-scope, but on Linux we need to know whether games will be launched via Proton or as a native Linux game, since we have different logic for both. As far as I'm aware, this information is only available through `%SteamFolder%/appcache/appinfo.vdf`, which contains information about what depots are installed and what operating system those depots are for. (if you're aware of a better way than this, please let me know, we previously used appmanifests for this by looking at platform sources and overrides, but that seems to no longer be a thing). However steamlocate currently doesn't touch this file, and its in a different format than the other files it currently deals with, so I'm not sure it fits here.

Might be more things? Those are the only two I can think of off the top of my head, otherwise we'd be happy to help just cleaning things up.

**Comment by CosmicHorrorDev** [comment]
Sure! All of those seem fine to me. The first one should be convenient for integration testing too, so that's a plus

Feel free to open another issue to track any potential changes you want, so that we can keep noise down on this one since there's probably at least a handful of people subscribed

Also timeline-wise I should be pretty close to finishing up v2.0, but I'm currently in the process of moving. We can always hash out any changes that would be a breaking change from the current design, but I'd imagine most of it could be done with non-semver breaking changes if required

**Comment by medzernik** [comment]
I noticed that the steamlocate Error can't be formatted using debug and therefore unwrapped or used pretty much anywhere, is that intended?

`steamlocate::Error` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`

**Comment by CosmicHorrorDev** [comment]
Good catch. Looks like I forgot to implement... well any traits for it. I'll add all the usual ones to it

**Comment by CosmicHorrorDev** [comment]
Back in for another quick update. There has been a lot of activity that hasn't been posted here, but it's not really worth rehashing. At this point there are only a few blockers before we can do the next release. I've added them to the tasklist in the original message

**Comment by CosmicHorrorDev** [comment]
I've gone ahead and published another alpha. This one is significantly different than the last, but the API should be more-or-less what to expect with the v2.0

The docs are all still out of date. I'm hoping to get that all sorted out and a beta prepared this weekend, but that depends on how much Lethal Company I wind up playing

**Comment by CosmicHorrorDev** [comment]
Went ahead and finished off the few remaining code changes and published a beta. All of the existing docs should be up-to-date now, but plenty of new stuff is still undocumented. Feel free to kick the tires on it while I finish writing up all the docs

https://docs.rs/steamlocate/2.0.0-beta.1/steamlocate/index.html

**Comment by CosmicHorrorDev** [comment]
Released another beta with a couple of fixes

- Fixes wasm support (did I mention we support wasm now?)
- Fixes apps failing to parse when the last updated field is cased `lastupdated` instead of `LastUpdated`

https://docs.rs/steamlocate/2.0.0-beta.2/steamlocate/index.html

**Comment by CosmicHorrorDev** [comment]
Published another beta

- Switches the `locate` feature to target-dependent hard dependencies
- `SteamDir::shortcuts()` now takes a `&self` instead of `&mut self`
- Lots of docs
- Lots of internal changes
- Other things that I don't remember

[`steamlocate@2.0.0-beta.4`](https://docs.rs/steamlocate/2.0.0-beta.4/steamlocate/index.html)

**Comment by CosmicHorrorDev** [comment]
Unless something major comes up this beta (maybe with some extra docs) will be published as the official v2.0 in a week or two

**Comment by CosmicHorrorDev** [comment]
[`steamlocate@2.0.0` is finally released :partying_face:](https://docs.rs/steamlocate/2.0.0/steamlocate/index.html)
